src/data/list/basic.lean

```diff
@@ -686,6 +695,46 @@ cons_head'_tail (head_mem_head' h)
 
 @[simp] theorem head'_map (f : α → β) (l) : head' (map f l) = (head' l).map f := by cases l; refl
 
+/-! ### Induction from the right -/
+
+/-- Induction principle from the right for lists: if a property holds for the empty list, and
+for `l ++ [a]` if it holds for `l`, then it holds for all lists. The principle is given for
+a `Sort`-valued predicate, i.e., it can also be used to construct data. -/
+@[elab_as_eliminator] def reverse_rec_on {C : list α → Sort*}
+  (l : list α) (H0 : C [])
+  (H1 : ∀ (l : list α) (a : α), C l → C (l ++ [a])) : C l :=
+begin
+  rw ← reverse_reverse l,
+  induction reverse l,
+  { exact H0 },
+  { rw reverse_cons, exact H1 _ _ ih }
+end
+
+/-- Bidirectional induction principle for lists: if a property holds for the empty list, the
+singleton list, and `a :: (l ++ [b])` from `l`, then it holds for all lists. This can be used to
+prove statements about palindromes. The principle is given for a `Sort`-valued predicate, i.e., it
+can also be used to construct data. -/
+def bidirectional_rec {C : list α → Sort*}
+    (H0 : C []) (H1 : ∀ (a : α), C [a])
+    (Hn : ∀ (a : α) (l : list α) (b : α), C l → C (a :: (l ++ [b]))) : ∀ l, C l
+| [] := H0
+| [a] := H1 a
+| (a :: b :: l) :=
+let l' := init (b :: l), b' := last (b :: l) (cons_ne_nil _ _) in
+have length l' < length (a :: b :: l), by { change _ < length l + 2, simp },
+begin
+  rw ←init_append_last (cons_ne_nil b l),
+  have : C l', from bidirectional_rec l',
+  exact Hn a l' b' ‹C l'›
+end
+using_well_founded { rel_tac := λ _ _, `[exact ⟨_, measure_wf list.length⟩] }
+
+/-- Like `bidirectional_rec`, but with the list parameter placed first for dot-notation. -/
```

Comment:
Do you have an example where `bidirectional_rec` doesn't work with dot notation? The notation shouldn't care where the argument appears, it will use the first argument with the right type. The `rec_on` variants are mainly around for more convenient partial applications.

Reply:
```suggestion
/-- Like `bidirectional_rec`, but with the list parameter placed first. -/
```

Reply:
Ah, I thought the argument had to appear first for dot notation... TIL. Thanks!